src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/OleVariant.cs

```diff
@@ -0,0 +1,560 @@
+            value.VarType = vt;
+            value.GetRawDataRef<T>() = (vt, sizeof(T)) switch
+            {
+                (VarEnum.VT_I1 or VarEnum.VT_UI1, 1) => rawValue,
+                (VarEnum.VT_I2 or VarEnum.VT_UI2 or VarEnum.VT_BOOL, 2) => rawValue,
+                (VarEnum.VT_ERROR or VarEnum.VT_HRESULT or VarEnum.VT_I4 or VarEnum.VT_UI4 or VarEnum.VT_R4 or VarEnum.VT_INT or VarEnum.VT_UINT, 4) => rawValue,
+                (VarEnum.VT_I8 or VarEnum.VT_UI8 or VarEnum.VT_R8 or VarEnum.VT_DATE, 8) => rawValue,
+                (VarEnum.VT_INT or VarEnum.VT_UINT or VarEnum.VT_UNKNOWN or VarEnum.VT_DISPATCH or VarEnum.VT_LPSTR or VarEnum.VT_BSTR or VarEnum.VT_LPWSTR or VarEnum.VT_SAFEARRAY
+                    or VarEnum.VT_CLSID or VarEnum.VT_STREAM or VarEnum.VT_STREAMED_OBJECT or VarEnum.VT_STORAGE or VarEnum.VT_STORED_OBJECT or VarEnum.VT_CF or VT_VERSIONED_STREAM, _) when sizeof(T) == nint.Size => rawValue,
+                (VarEnum.VT_CY or VarEnum.VT_FILETIME, 8) => rawValue,
+                (VarEnum.VT_RECORD, _) when sizeof(T) == sizeof(Record) => rawValue,
+                _ when vt.HasFlag(VarEnum.VT_BYREF) && sizeof(T) == nint.Size => rawValue,
+                _ when vt.HasFlag(VarEnum.VT_VECTOR) && sizeof(T) == sizeof(Vector<byte>) => rawValue,
+                _ when vt.HasFlag(VarEnum.VT_ARRAY) && sizeof(T) == nint.Size => rawValue,
+                (VarEnum.VT_BLOB or VarEnum.VT_BLOB_OBJECT, _) when sizeof(T) == sizeof(Blob) => rawValue,
+                _ => throw new ArgumentException("Size of T should be the same size as the value specified by vt")
+            };
+
+            return value;
+        }
+
+        /// <summary>
+        /// A <see cref="OleVariant"/> instance that represents a null value with <see cref="VarEnum.VT_NULL"/> type.
+        /// </summary>
+        public static OleVariant Null { get; } = new() { VarType = VarEnum.VT_NULL };
+
+        private readonly void ThrowIfNotVarType(params VarEnum[] requiredType)
+        {
+            if (Array.IndexOf(requiredType, VarType) == -1)
+            {
+                throw new InvalidOperationException($"Variant type {VarType} cannot be cast to any of the supported variant types: [{string.Join(", ", requiredType)}]");
+            }
+        }
+
+#pragma warning disable CS0618 // Type or member is obsolete
+        /// <summary>
+        /// Create a managed value based on the value in the <see cref="OleVariant"/> instance.
+        /// </summary>
+        /// <typeparam name="T">The managed type to create an instance of.</typeparam>
+        /// <returns>The managed value contained in this variant.</returns>
+        /// <exception cref="ArgumentException">When <typeparamref name="T"/> does not match the <see cref="VarType"/> of the <see cref="OleVariant"/>.</exception>
+        public readonly T As<T>()
+        {
+            if (VarType == VarEnum.VT_EMPTY)
+            {
+                return default!;
```

Comment:
Why is the `!` needed?

Reply:
The return type is non-nullable, but `default` can be `null`. I'll update the API to return a nullable value.

Reply:
Because `T` isn't constrained, right?

Reply:
Yes. I changed the return type to `T?` (which is basically "T" but if it's a reference type it can be `null`) to be more accurate.